components/sup/src/manager/service/spec.rs

```diff
@@ -266,6 +266,113 @@ impl ServiceSpec {
+        use DesiredState::{Down,
+                           Up};
+        match (old.as_ref().map(|o| o.desired_state),
+               old,
+               new.as_ref().map(|n| n.desired_state),
+               new)
+        {
+            // theoretically shouldn't happen, but no harm if it does.
+            (_, None, _, None)
+            // Somebody manually added a spec file that for some
+            // reason stated the service should be down. Weird, but
+            // okay....
+            | (_, None, Some(Down), Some(_))
+            // A stopped service's spec file was removed
+            | (Some(Down), Some(_), _, None)
+            // A stopped service's spec file was changed, but it is
+            // still supposed to be down. This would also likely
+            // require manual intervention.
+            | (Some(Down), Some(_), Some(Down), Some(_)) => {
+                // None of these situations require us to do anything
+                // in the way of starting, stopping, restarting, or
+                // modifying a service.
+                None
+            }
+
+            // A running service's spec file was removed (e.g., hab
+            // svc unload)
+            (Some(Up), Some(old), _, None)
+            // A running service was told to stop (e.g., hab svc stop)
+            | (Some(Up), Some(old), Some(Down), Some(_)) => {
+                Some(ServiceOperation::Stop(old))
+            }
+
+            // A new spec file was added (e.g., hab svc load)
+            (_, None, Some(Up), Some(new))
+            // A previously stopped service was started (e.g., hab svc start)
+            | (Some(Down), Some(_), Some(Up), Some(new)) => {
+                Some(ServiceOperation::Start(new))
+            }
+
+            // The configuration of a running service was somehow changed
+            (Some(Up), Some(running_spec), Some(Up), Some(disk_spec)) => {
+                if running_spec == disk_spec {
+                    // Given how this function is called, this is
+                    // unlikely to happen, but if it does, we don't
+                    // have to do anything.
+                    None
+                } else {
+                    // TODO (CM): In the future, this would be the
+                    // place where we can evaluate what has changed
+                    // between the spec-on-disk and our in-memory
+                    // representation and potentially just bring our
+                    // in-memory representation in line without having
+                    // to restart the entire service.
+                    debug!("Reconciliation: '{}' queued for restart",
+                           running_spec.ident);
+                    Some(ServiceOperation::Restart { to_stop:  running_spec,
+                                                     to_start: disk_spec, })
+                }
+            }
+
+            // These are unreachable because a None spec can't yield a
```

Comment:
Instead of matching on `(Option<DesiredState>, Option<ServiceSpec>, Option<DesiredState>, Option<ServiceSpec>)` could you match on `(Option<(DesiredState, ServiceSpec)>, Option<DesiredState, ServiceSpec>)` to remove thes invalid states?

Reply:
Good call... that's a much clearer formulation! :+1: